File: homecloud_core/signing.py
"""HomeCloud request signing (SigV1) — internal only."""

from __future__ import annotations

import hashlib
import hmac
from datetime import datetime, timezone


def build_string_to_sign(
    *,
    method: str,
    path: str,
    timestamp: str,
    account_id: str,
) -> str:
    return f"{method.upper()}\n{path}\n{timestamp}\n{account_id}"


def compute_signature(*, secret: str, string_to_sign: str) -> str:
    return hmac.new(
        secret.encode("utf-8"),
        string_to_sign.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()


def sign_request_headers(
    *,
    access_key_id: str,
    secret: str,
    method: str,
    path: str,
    account_id: str,
    timestamp: datetime | None = None,
) -> dict[str, str]:
    ts = (timestamp or datetime.now(timezone.utc)).replace(microsecond=0)
    ts_str = ts.isoformat().replace("+00:00", "Z")
    string_to_sign = build_string_to_sign(
        method=method,
        path=path,
        timestamp=ts_str,
        account_id=account_id,
    )
    signature = compute_signature(secret=secret, string_to_sign=string_to_sign)
    return {
        "X-Homecloud-Access-Key-Id": access_key_id,
        "X-Homecloud-Date": ts_str,
        "X-Homecloud-Signature": signature,
    }

File: java/src/main/java/com/homecloudlab/sdk/Signing.java
package com.homecloudlab.sdk;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;

/**
 * SigV1 HMAC-SHA256. Canonical string: {@code {METHOD}\n{path}\n{timestamp}\n{account_id}}.
 */
public final class Signing {
    public static final String HEADER_ACCESS_KEY_ID = "X-Homecloud-Access-Key-Id";
    public static final String HEADER_DATE = "X-Homecloud-Date";
    public static final String HEADER_SIGNATURE = "X-Homecloud-Signature";
    public static final String HEADER_SESSION_TOKEN = "X-Homecloud-Session-Token";
    public static final String HEADER_IDEMPOTENCY = "Idempotency-Key";

    private static final DateTimeFormatter TS =
            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'").withZone(ZoneOffset.UTC);

    private Signing() {}

    public static String buildStringToSign(String method, String path, String timestamp, String accountId) {
        return method.toUpperCase(Locale.ROOT) + "\n" + path + "\n" + timestamp + "\n" + accountId;
    }

    public static String computeSignature(String secret, String stringToSign) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            byte[] raw = mac.doFinal(stringToSign.getBytes(StandardCharsets.UTF_8));
            StringBuilder hex = new StringBuilder(raw.length * 2);
            for (byte b : raw) {
                hex.append(String.format("%02x", b));
            }
            return hex.toString();
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            throw new HomeCloudException("HMAC-SHA256 failed", e);
        }
    }

    public static String formatTimestamp(Instant instant) {
        return TS.format(instant.truncatedTo(ChronoUnit.SECONDS));
    }

    static Map<String, String> signHeaders(
            String accessKeyId,
            String secret,
            String method,
            String path,
            String accountId,
            Instant now,
            String sessionToken) {
        String ts = formatTimestamp(now);
        String sig = computeSignature(secret, buildStringToSign(method, path, ts, accountId));
        Map<String, String> h = new LinkedHashMap<>();
        h.put(HEADER_ACCESS_KEY_ID, accessKeyId);
        h.put(HEADER_DATE, ts);
        h.put(HEADER_SIGNATURE, sig);
        if (sessionToken != null && !sessionToken.isBlank()) {
            h.put(HEADER_SESSION_TOKEN, sessionToken);
        }
        return h;
    }
}
